### DISCUSSION TITLE
How to use "--extract-media" and preserve regular (non-SHA1) filenames?

### DISCUSSION THREAD

**Discussion by johnmapeson** [jgm/pandoc#9697] [Q&A] [answered]
I try to download an HTML page from the internet as Markdown and with all the linked images. Here is what I use currently for this:

```
pandoc http://foo.bar/baz.html -o baz.md --wrap=none --extract-media="./images"
```

The problem is that the images are saved with SHA1-based filenames, and even though I've read the user's guide,

> `--extract-media=DIR`
>
> Extract images and other media contained in or linked from the source document to the path DIR, creating it if necessary, and adjust the images references in the document so they point to the extracted files. Media are downloaded, read from the file system, or extracted from a binary container (e.g. docx), as needed. The original file paths are used if they are relative paths not containing `..`. Otherwise filenames are constructed from the SHA1 hash of the contents.

I still don't understand how to preserve regular filenames. What exactly should I do to preserve regular filenames, like `foo.png`, instead of having SHA1-based ones, like `6f10f22599031f605714f6b3f1153b5c12d96084.png`?

**Comment by jgm** [comment, marked as answer]
See commit eddedbfc14916aa06fc01ff04b38aeb30ae2e625
I am guessing that on the HTML page, some percent-encoding is used in the image paths.  We use SHA1 in that case to avoid a security vulnerability.

**Reply by johnmapeson** [reply]
Yes, you are correct, there is percent-encoding there. Actually I would prefer to reduce security temporary, because I trust this website and there are about 200 images there, but most probably this is not possible.

**Reply by tarleb** [reply]
You could use a Lua filter like the below.

``` lua
local hex_to_char = function(x)
  return string.char(tonumber(x, 16))
end

local unescape = function(url)
  return url:gsub("%%(%x%x)", hex_to_char)
end

function Image (img)
  -- ignore local images
  if not img.src:match('^https://') then
    return nil
  end

  local mt, contents = pandoc.mediabag.fetch(img.src)
  img.src = pandoc.path.filename(unescape(img.src))
  pandoc.mediabag.insert(img.src, mt, contents)

  return img
end
```

**Reply by johnmapeson** [reply]
Tarleb, many thanks!

**Reply by johnmapeson** [reply]
Albert, hello! Maybe this is too much, but could you show how to adjust this filter so that it will handle images such as `../image.png` as well, that is, to preserve the original filename of image if there is `..` in its path?

**Comment by tarleb** [comment]
Changing the line

    if not img.src:match('^https://') then

to

    if not img.src:match('^https://') or img.src:match '%.%.' then

should do the trick.

**Reply by johnmapeson** [reply]
Hmm, no, this haven't helped. I have also tried to enclose `'%.%.'` in `()`, that is,

``` lua
if not img.src:match('^https://') or img.src:match('%.%.') then
```

but this doesn't seem to make any difference.

**Reply by tarleb** [reply]
Ah yes, my bad. It should either be `and not` instead of `or`, or everything between `not` and `then` should be placed in parentheses.

**Reply by johnmapeson** [reply]
This works, many thanks! :)
